### DISCUSSION TITLE
v2.1.0

### DISCUSSION THREAD

**Discussion by alexdelprete** [alexdelprete/ha-abb-powerone-pvi-sunspec#35] [Announcements]
- Adapted code to pymodbus latest version (v3.0.0)
- Added some unknown inverter models

<hr /><em>This discussion was created from the release <a href='https://github.com/alexdelprete/ha-abb-powerone-pvi-sunspec/releases/tag/v2.1.0'>v2.1.0</a>.</em>

**Comment by dexmorgan8888** [comment]
@alexdelprete I was on V2.1.0 and my setup was working well.

I wanted to change the refresh interval. As there was no config change UI, I attemped to delete the integration and recreate.

But when I recreated, it said successful although there was no device created. 
![image](https://user-images.githubusercontent.com/18218130/201711830-d61e2368-4006-48ac-bfc5-8393bd1b2d10.png)

There are 2 entries in the log
![image](https://user-images.githubusercontent.com/18218130/201712049-344cd6dc-4deb-48a1-89d6-77da1cfa1ac3.png)

```
This error originated from a custom integration.

Logger: custom_components.abb_powerone_pvi_sunspec
Source: custom_components/abb_powerone_pvi_sunspec/__init__.py:193
Integration: ABB Power-One PVI SunSpec (documentation, issues)
First occurred: 00:17:33 (1 occurrences)
Last logged: 00:17:33

(opt_comm_model) Model unknown, report to @alexdelprete on the forum the following data: Manuf.:  - Model:  - Options:  - OptModel:  - OptModelInt: 0

```

and

```
Logger: homeassistant.components.sensor
Source: custom_components/abb_powerone_pvi_sunspec/__init__.py:319
Integration: Sensor (documentation, issues)
First occurred: 00:17:33 (1 occurrences)
Last logged: 00:17:33

Error while setting up abb_powerone_pvi_sunspec platform for sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 281, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/abb_powerone_pvi_sunspec/sensor.py", line 21, in async_setup_entry
    hub.read_sunspec_modbus_data()
  File "/config/custom_components/abb_powerone_pvi_sunspec/__init__.py", line 193, in read_sunspec_modbus_data
    return self.read_sunspec_modbus_model_1() and self.read_sunspec_modbus_model_101_103() and self.read_sunspec_modbus_model_160()
  File "/config/custom_components/abb_powerone_pvi_sunspec/__init__.py", line 319, in read_sunspec_modbus_model_101_103
    _LOGGER.debug("(read_rt_1) Inverter Type (str): %s", INVERTER_TYPE[invtype])
KeyError: 0

```

I have also downgraded to 2.0.11 and 2.0.9 (this was my first install that worked) with HA restart each time.
I also deleted the custom_components\abb_powerone_pvi_sunspec each time.
All didn't work.

**Reply by alexdelprete** [reply]
that means it's not installing correctly, it's not a component issue, but something is preventing a correct installation of the component.

describe the HA environment: core, supervised, etc. what version. what hardware. did you change something related to HA lately? python etc.

**Reply by dexmorgan8888** [reply]
@alexdelprete Thanks for looking into this!

Installation is on HA OS on Proxmox VM.
Versions
Home Assistant 2022.11.2
Supervisor 2022.10.2
Operating System 9.3
Frontend 20221108.0 - latest

Hardware is a HP EliteDesk 800 with 4 x Intel(R) Core(TM) i5-4590 CPU @ 3.30GHz
Using 1 core for HA OS. This setup has been running without any issues.

I don't think I changed anything. The whole thing was working perfectly well until I reinstalled the integration.

Did I miss anything? Perhaps something to clean before installation? btw, I am reusing the previous name, not sure if that affects anything.

**Comment by alexdelprete** [comment]
The only prerequirement is pymodbus library 3.0, but that should be pulled automatically. Do you have other components installed using modbus?

So before removing the component, you didn't change anything? Did you reinstall HA?

When you configure it, are you sure you are using the correct parameters? IP, slave id, etc. Show me the screenshot of the configuration before submitting it.

**Comment by dexmorgan8888** [comment]
No, didn't change anything, didn't reinstall either.
I don't use other modbus component, i dont think.

![image](https://user-images.githubusercontent.com/18218130/201737555-43ccf300-1e3d-43fd-9f56-0e4ef4df859f.png)

I only changed the name and IP address. Everything else is default.
I also tried telnet ip and port 502 to ensure port was open. It was.

**Comment by alexdelprete** [comment]
1. What happens right after you hit SUBMIT on the config panel?
2. Are you sure the IP is correct?
3. Are you sure the slave address 247 is correct?
4. Telnet doesn't mean anything, you need to do a modbus query, using a specific tool (qmodmaster, etc.) to see if the inverter is answering appropriately to modbus queries, because for example if the slave address is wrong, telnet tells you that port 502 is open, but then the query to the inverter would fail.

**Comment by dexmorgan8888** [comment]
1. After a few seconds of working, a pop up appears saying installation is successful. The ABB icon is created, but without device and entities. The device and entities tabs say that none of the device and entities were created. The log shows error.
2. Yes. I use the same IP to access internal inverter webserver.
3. No idea. Where do I get this? I searched the user manual for 'slave' without luck.
4. Ok, I can explore with your guidance.

In the meantime, I have restored HA from my backup and everything is back to normal.
Can I check the slave address in this setup? how?
Can I also change the refresh interval manually? 

Thanks!

**Reply by alexdelprete** [reply]
> No idea. Where do I get this? I searched the user manual for 'slave' without luck.

I'm fairly confident you are using the wrong Slave ID (also called Unit ID). ModBus is a bus, and you can have multiple devices on the bus, the inverter is one of the modbus devices, in order to communicate with a ModBus device, you need to know its modbus slave id otherwise he would ignore the commands sent on the bus.

You can check the Slave/Unit ID on the inverter's web interface, or on the LCD panel on the inverter too.

You can read the config in `/config/.storage/core.config_entries` (be careful with this file, it's HA registry, if you make mistakes HA won't startup, so save it before editing it) search for `abb_powerone_pvi_sunspec`, you should find this section with all the settings:

```json
      {
        "entry_id": "669ea7e7feb9b859d9b42bdf729aa64b",
        "version": 1,
        "domain": "abb_powerone_pvi_sunspec",
        "title": "ABB VSN300",
        "data": {
          "name": "ABB VSN300",
          "host": "abb-vsn300.axel.dom",
          "port": 502,
          "slave_id": 2,
          "base_addr": 0,
          "scan_interval": 30
        },
```

Once you find it, please take note of the `slave_id` and `base_addr`. You can also modify the `scan_interval`, and after saving you need to restart HA.

Let me know how it goes.
